Feedback #7 #8

Blauben opened the pull request.

schuhmaj commented:
FYI: I am writing this here for quick reference. The changes which I pulled in from the `upstream` repository embed useful version information into the program, accessible via `#include "polyhedralGravity/Info.h"`

(The same holds for the Python Implementation)

You may use this for benchmarking to e.g. print out the commit and extract it automatically to not lose track over measurements across different versions.

Blauben commented:
I have addressed the feedback and completed the parallelization using TBB via thrust. The python interface is still broken, I will look into that at a later date.

schuhmaj commented on script/__pycache__/scale_up_mesh.cpython-313.pyc:
Try to never commit binary files!

schuhmaj commented on CMakeLists.txt:
```diff
@@ -42,8 +42,12 @@ cmake_dependent_option(BUILD_POLYHEDRAL_GRAVITY_LIBRARY "Builds the library (Def
+# Option to build benchmarks or not
+option(BUILD_POLYHEDRAL_GRAVITY_TIME_MEASUREMENT "Set to on if the benchmark executable should be built (Default: ON)" ON)
+# Option to enable Include-What-You-Use warnings during compilation
+option(ENABLE_IWYU "Set to on to enable Include-What-You-Use warnings (Default: OFF)" OFF)
```
Good idea

schuhmaj commented on example-config/example_eros_scaled_1.yaml:
```diff
@@ -1,11 +0,0 @@
----
```
it might be a good idea to check these large files into Git LFS (I mean the `.node` and `.face` files)

schuhmaj commented on script/annotate_iwyu.py:
```diff
@@ -0,0 +1,34 @@
+    return None
+
+def main(iwyu_log):
+    current_file = None
```
If you put a *script*  into the repository, you also need to document it. You can put a docstring `"""On top of the file""`. This is the usual practice for Python scripts.

schuhmaj commented on script/scale_mesh.py:
```diff
@@ -0,0 +1,148 @@
+            f.write(line)
+
+
+def main():
```
Same here. If the script is not only "throw-away", but checked into the repository, please document its functionality. Here, also the indivudal functions require some documentation.

schuhmaj commented on src/polyhedralGravity/model/KDTree/plane_selection/LogNPlane.cpp:
```diff
@@ -112,16 +60,16 @@ namespace polyhedralGravity {
         result.reserve(events.size() / 6);
         //preparing the map by initializing all faces with them having area in both sub bounding boxes
-        std::for_each(events.cbegin(), events.cend(), [&result](const auto &event) {
+        thrust::for_each(thrust::host, events.cbegin(), events.cend(), [&result](const auto &event) {
```
In our use case `thrust::host` is always `CPP`, i.e. serial execution. You can also omit this argument.

schuhmaj commented on src/polyhedralGravity/model/KDTree/plane_selection/LogNPlane.cpp:
```diff
@@ -182,10 +135,10 @@ namespace polyhedralGravity {
 
     //Step 4
-    std::unique_ptr<PlaneEventList> LogNPlane::mergePlaneEventLists(const PlaneEventList &first, const PlaneEventList &second) {
+    std::unique_ptr<PlaneEventVector> LogNPlane::mergePlaneEventLists(const PlaneEventVector &first, const PlaneEventVector &second) {
```
"Täglich grüßt das Murmeltier" - Maybe an overall line next to "Step X", what is actually happening in that respective step

schuhmaj commented on CMakeLists.txt:
```diff
@@ -42,8 +42,12 @@ cmake_dependent_option(BUILD_POLYHEDRAL_GRAVITY_LIBRARY "Builds the library (Def
 # Option to build tests or not
 option(BUILD_POLYHEDRAL_GRAVITY_TESTS "Set to on if the tests should be built (Default: ON)" ON)
+# Option to build benchmarks or not
+option(BUILD_POLYHEDRAL_GRAVITY_TIME_MEASUREMENT "Set to on if the benchmark executable should be built (Default: ON)" ON)
```
Please set this to `OFF` by default. This causes the build failure of the Python interface

schuhmaj reviewed (approved):
# Feedback

## Points to improve

I don't really have anything to complain about these changes. I've added few minor remarks, but mostly I can give you the following overall feedback:

- ⚠️Please avoid the committing of binary files. E.g. `script/__pycache__/*` - These files need to be git ignored!
- ☑️If the file is multiple megabytes large, it makes sense to commit it via `git lfs`. Here, I wouldn't recommend changing anything yet since I have want to re-organize our `ressource` folders anyway (duplicate information between `script/mesh` and `example-config/data`)
- ⚠️ The Python scripts are not yet documented properly but checked in. To avoid throw-away-use-once script and foster re-use, please document their functionality
- ❗️Please fix the test cases!
  - ‼️ Given the fact that the Windows C/C++ test suit works, I assume the failing test on Linux/ macOS is only caused by numerical rounding errors. Nevertheless, this needs to be investigated
  - For the Python Interface tests, please take a look at the logs [here](https://github.com/Blauben/polyhedral-gravity-model/runs/36457533041#step:5:553). You are building your time measurement main method along the Python interface which cannot work since the Python interface does not include all dependencies of the C++ library. Hence, set `BUILD_POLYHEDRAL_GRAVITY_TIME_MEASUREMENT` in `CMakeLists.txt` by default to `OFF`. 
- ⚠️ You have deleted the `main.cpp` - probably by accident

----
Legend:
❗️Definitly fix
⚠️ Please fix if you find time
☑️ No task for you/ I will fix


## Overall Verdict

Independent of the bits above:The final codebase is very nice and well-organized 👍. 
Well done!

Blauben commented:
Thanks for the valuable feedback, I've integrated the requested changes. Finally the python interface works again and I've been able to fix the last bugs revealed by the regression tests.
